### DISCUSSION TITLE
can not uninstall

### DISCUSSION THREAD

**Discussion by pas059** [wixtoolset/issues#8025] [Questions] [answered]
Hi,
on some machine, my installer does not uninstall. Even using a command line like 'msiexec /x ffff.msi /l*v ffff.log'. The only way to uninstall is to restaure the system!

Someone can help?

**Comment by pas059** [comment]
Hi,
after exploring the .log i suspect a problem during driver uninstall. Below a part of the log:
`MSI (s) (60:A4) [18:49:46:414]: Executing op: ActionStart(Name=UninstallUsbXpress,,)
MSI (s) (60:A4) [18:49:46:414]: Executing op: CustomActionSchedule(Action=UninstallUsbXpress,ActionType=3073,Source=BinaryData,Target=WixQuietExec,CustomActionData="pnputil.exe" /delete-driver "C:\ProgramData\Bxxxxxxx\Mxxxx\Driver USBXpress 6.7\siusbxp.inf" /uninstall)
MSI (s) (60:14) [18:49:46:415]: Invoking remote custom action. DLL: C:\WINDOWS\Installer\MSIF50B.tmp, Entrypoint: WixQuietExec
MSI (s) (60:40) [18:49:46:415]: Generating random cookie.
MSI (s) (60:40) [18:49:46:416]: System Environment : Client has accepted UAC prompt
MSI (s) (60:40) [18:49:46:416]: System Environment : Client has accepted UAC prompt
MSI (s) (60:40) [18:49:46:418]: Created Custom Action Server with PID 7736 (0x1E38).
MSI (s) (60:70) [18:49:46:429]: Running as a service.
MSI (s) (60:70) [18:49:46:431]: Hello, I'm your 64bit Elevated Non-remapped custom action server.
WixQuietExec:  Utilitaire Plug-and-Play Microsoft
WixQuietExec:  
WixQuietExec:  Impossible de désinstaller le package de pilotes : Le fichier spécifié n’est pas un INF OEM installé.
WixQuietExec:  Error 0xe000023c: Command line returned an error.
WixQuietExec:  Error 0xe000023c: QuietExec Failed
WixQuietExec:  Error 0xe000023c: Failed in ExecCommon method
CustomAction UninstallUsbXpress returned actual error code 1603 (note this may not be 100% accurate if translation happened inside sandbox)
`
uninstalling works on my machines but not on some customer machine.
Someone has an idea?

regards

**Comment by chrpai** [comment, marked as answer]
The custom action is throwing an exception and that is rolling back your uninstall.  To resolve for this one machine you need to:

a. Fix the reason for 0xe000023c.  Google search says this is an error occurred while uninstalling driver package.  Fixing this may prove hard.

b. Modify the cached installer to mark the custom action's return processing to ignore errors.  This can be done by modifying your built MSI with ORCA and adding decimal 64 (bitmask 0x00000040 logical or) to the custom action type column.    Save the MSI and recache it into C:\windows\installer  but running the command   `msiexec /i foo.msi REINSTALL=ALL REINSTALLMODE=vomus`.   After the reinstall is complete attempt uninstall again and it should get past this error.

c. Rebuild the machine.


Note:  This is a good reason to limit the use of custom actions, always set uninstall custom actions to continue/ignore so they run as 'best effort' and never block uninstall,  always test thoroughly on virtual machines not your own machine.  If this MSI has made it to production be aware that you have a ticking time bomb with your customers so consider making the updated MSI available and/or writing a KB article.  You will want to make this change for your next release.

https://learn.microsoft.com/en-us/windows/win32/msi/custom-action-return-processing-options

**Comment by pas059** [comment]
Hello,
thank you very much Christopher for your reply and advices.
I followed your méthod to modify the current msi (with orca), reinstall the modified msi (by command line) and finally uninstall, and this uninstalls, yeah!! the machine seems clean without restoring.
For the next release, i changed the `Return="check"` attribute of the uninstall CustomAction element by `Return="ignore"`.

Since I haven't found yet why the driver doesn't uninstall, i want to unsellect this feature by default. But i did not find a way to do this; currently i do this:
```
		<Feature Id="DriverFeature" Title="!(loc.DriverFeatures)" Level="1" AllowAbsent="yes" Display="expand">
			<ComponentRef Id="Driver_USBXpress_Files" />
			<ComponentGroupRef Id="Driver_USBXpress_67" />
			<Feature Id="Driver_USBXpress_67_Feature" Title="!(loc.Driver_USBXpress_67_Feature)" Level="5" AllowAbsent="yes" />
		</Feature>
```
but the Driver_USBXpress_67_Feature is always selected despite the level (5). 
What is wrong?

regards

**Comment by chrpai** [comment]
What is your INSTALLLEVEL property set to?  Is this a first time install or an upgrade? Have you looked in the MSI log for clues?

**Comment by pas059** [comment]
Hi,
i tried with different values and this had no influence ... but on my machine for which this is an upgrade of course. But when trying on a new machine, for which it's a first time install, this works! hence the usefulness of your second questions Christian. Thanks again

**Comment by pas059** [comment]
Hi,

i tested successfully another method for uninstall my app installed with installer that fails to uninstall (ouf). Making an installer for an newer version that, of course, correct the initial problem (by changing the `Return="check"` attribute of the uninstall CustomAction element by `Return="ignore"`). I think tha this will be more acceptable from an end user user point of view.

regards
